provider/pkg/metadata/naming.go

```diff
@@ -17,17 +17,27 @@ package metadata
 import (
 	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
 // AssignNameIfAutonamable generates a name for an object. Uses DNS-1123-compliant characters.
 // All auto-named resources get the annotation `pulumi.com/autonamed` for tooling purposes.
-func AssignNameIfAutonamable(randomSeed []byte, obj *unstructured.Unstructured, propMap resource.PropertyMap, urn resource.URN) {
+func AssignNameIfAutonamable(randomSeed []byte, engineAutonaming *pulumirpc.CheckRequest_AutonamingOptions,
+	obj *unstructured.Unstructured, propMap resource.PropertyMap, urn resource.URN) {
 	contract.Assertf(urn.Name() != "", "expected non-empty name in URN: %s", urn)
 	if !IsNamed(obj, propMap) && !IsGenerateName(obj, propMap) {
 		prefix := urn.Name() + "-"
 		autoname, err := resource.NewUniqueName(randomSeed, prefix, 0, 0, nil)
 		contract.AssertNoErrorf(err, "unexpected error while creating NewUniqueName")
+		if engineAutonaming != nil {
+			switch engineAutonaming.Mode {
+			case pulumirpc.CheckRequest_AutonamingOptions_DISABLE:
+				return
+			case pulumirpc.CheckRequest_AutonamingOptions_ENFORCE, pulumirpc.CheckRequest_AutonamingOptions_PROPOSE:
+				autoname = engineAutonaming.ProposedName
+			}
```

Comment:
We should also probably validate the proposed name, since the user could specify a template that isn't valid for k8s.

I _think_ that would be [IsDNS1123Subdomain](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation#IsDNS1123Subdomain) but @rquitales should double check this! 

https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names

Reply:
Do we validate `urn.Name()` anywhere?

Reply:
Yes, [IsDNS1123Subdomain](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation#IsDNS1123Subdomain) is the correct validator.

Reply:
> Do we validate `urn.Name()` anywhere?

That's a good question, I don't think we do! I guess we offload all of that to the cluster, so maybe we don't need to validate anything here.

I'm curious what the user experience looks like in that situation? If it's pretty clear how to resolve the issue then this is probably fine.

Reply:
This is the error that one gets in both cases:

> error: resource "urn:pulumi:dev::kubernetes-ts::kubernetes:apps/v1:Deployment::nginx!!!!" was not successfully created by the Kubernetes API server: Deployment.apps "nginx!!!!-6626ce94" is invalid: metadata.name: Invalid value: "nginx!!!!-6626ce94": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')